src/storage/http/StorageHttpDownloadHandler.cpp

```diff
@@ -144,46 +190,56 @@ void StorageHttpDownloadHandler::onError(ProxygenError error) noexcept {
                << proxygen::getErrorString(error);
 }
 
-bool StorageHttpDownloadHandler::downloadSSTFiles(const std::string& hdfsHost,
-                                                  int32_t hdfsPort,
-                                                  const std::string& hdfsPath,
-                                                  const std::vector<std::string>& parts) {
-    static std::atomic_flag isRunning = ATOMIC_FLAG_INIT;
-    if (isRunning.test_and_set()) {
-        LOG(ERROR) << "Download is not completed";
-        return false;
-    }
-
+bool StorageHttpDownloadHandler::downloadSSTFiles() {
     std::vector<folly::SemiFuture<bool>> futures;
-
-    for (auto& part : parts) {
+    for (auto& part : parts_) {
         PartitionID partId;
         try {
             partId = folly::to<PartitionID>(part);
         } catch (const std::exception& ex) {
-            isRunning.clear();
             LOG(ERROR) << "Invalid part: \"" << part << "\"";
             return false;
         }
-
-        auto downloader = [hdfsHost, hdfsPort, hdfsPath, partId, this]() {
-            auto hdfsPartPath = folly::stringPrintf("%s/%d", hdfsPath.c_str(), partId);
-            auto partResult = kvstore_->part(spaceID_, partId);
-            if (!ok(partResult)) {
-                LOG(ERROR) << "Can't found space: " << spaceID_ << ", part: " << partId;
+        auto hdfsPartPath = folly::stringPrintf("%s/%d", hdfsPath_.c_str(), partId);
+        auto existStatus = helper_->exist(hdfsHost_, hdfsPort_, hdfsPartPath);
+        if (!existStatus.ok()) {
+            LOG(ERROR) << "Run Hdfs Test failed. " << existStatus.status().toString();
+            return false;
+        }
+        bool exist = existStatus.value();
+        if (!exist) {
+            LOG(WARNING) << "Hdfs path non exist. hdfs://" << hdfsHost_ << ":" << hdfsPort_ << hdfsPartPath;
+            continue;
+        }
+        auto partResult = kvstore_->part(spaceID_, partId);
+        if (!ok(partResult)) {
+            LOG(ERROR) << "Can't found space: " << spaceID_ << ", part: " << partId;
+            return false;
+        }
+        auto partDataRoot = value(partResult)->engine()->getDataRoot();
+        std::string localPath;
+        if (edge_.has_value()) {
+            localPath = folly::stringPrintf(
+                "%s/download/edge/%d", partDataRoot, edge_.value());
```

Comment:
You put all parts data together?

Reply:
// output dir:
    // -- 1/1-1.sst
    //     /1-2.sst
    //     /1-3.sst
    // -- 2/2-1.sst
    //     /2-2.sst

Reply:
Got it.